### PR TITLE
generic component to highlight any part of the ui

### PR DESCRIPTION
**Pull request by nemesis09**
**Fixes:**
https://issues.redhat.com/browse/ODC-4999

**Description:**
Create a generic component to highlight any part of UI with animation

**Screens:**
_highlight component behaviour ideal_
Note: three components are highlighted simultaneously for illustration only.
quickstarts_highlight

_highlight component behaviour when clicked anywhere on the ui after animation has started_
stophighlightanimation

_highlight component behaviour autoscroll to bring component to highlight in viewport if it is hidden_
highlightautoscroll

_guided tour behaviour unchanged with updated `Spotlight` component_
guidedtour_highlight

**Test Coverage:**
Screenshot from 2020-11-11 01-37-50

**Browser Conformance:**
- [x] Firefox
- [ ] Chrome
- [ ] Safari
- [ ] Edge

### CODE REVIEW

File: frontend/packages/console-shared/src/components/highlight/Highlight.tsx
Comment (rohitkrai03) [inline frontend/packages/console-shared/src/components/highlight/Highlight.tsx:19]: The implementation is exactly same as the Spotlight component from guided tour. Instead of creating a new one you should just move that one to console-shared and add some customization for this use case.

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]

File: frontend/packages/console-shared/src/components/highlight/Highlight.tsx
Comment (nemesis09) [reply frontend/packages/console-shared/src/components/highlight/Highlight.tsx]: updated with customized spotlight component

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx:30]: DRY. Store this component in a variable and render conditionally.

**Comment by lwrigh** [comment]
@nemesis09 Here are the animation specifications from the final design doc: 

Blink animation is triggered by click/tap initiating step
Blink animation is delayed start by 0.5 seconds ⬅️
Blink animation consists of (all at the same time and over 1.2 sec duration)
Scale: shape increases size from 100% to 150% ⬅️
Border width: from 4px to 8px ⬅️
Opacity: from 100% to 0%

I took a closer look at the GIF and CSS and I think the elements that are missing are the specifications with the arrows next to them. Let me know if you need me to take a closer look at any animation specifications. 

Here's the link to the final design doc if you need to reference it: https://docs.google.com/document/d/1T41IzREwgfqR6NVpXI8e4psfkBR39gWt0eLfvA9pkGs/edit?usp=sharing

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx
Comment (nemesis09) [reply frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx]: updated using variable.

**Comment by nemesis09** [comment]
> @nemesis09 Here are the animation specifications from the final design doc:
> 
> Blink animation is triggered by click/tap initiating step
> Blink animation is delayed start by 0.5 seconds arrow_left
> Blink animation consists of (all at the same time and over 1.2 sec duration)
> Scale: shape increases size from 100% to 150% arrow_left
> Border width: from 4px to 8px arrow_left
> Opacity: from 100% to 0%
> 
> I took a closer look at the GIF and CSS and I think the elements that are missing are the specifications with the arrows next to them. Let me know if you need me to take a closer look at any animation specifications.
> 
> Here's the link to the final design doc if you need to reference it: https://docs.google.com/document/d/1T41IzREwgfqR6NVpXI8e4psfkBR39gWt0eLfvA9pkGs/edit?usp=sharing

updated animation addressing the suggested requirements. PTAL

**Comment by lwrigh** [comment]
> > @nemesis09 Here are the animation specifications from the final design doc:
> > Blink animation is triggered by click/tap initiating step
> > Blink animation is delayed start by 0.5 seconds arrow_left
> > Blink animation consists of (all at the same time and over 1.2 sec duration)
> > Scale: shape increases size from 100% to 150% arrow_left
> > Border width: from 4px to 8px arrow_left
> > Opacity: from 100% to 0%
> > I took a closer look at the GIF and CSS and I think the elements that are missing are the specifications with the arrows next to them. Let me know if you need me to take a closer look at any animation specifications.
> > Here's the link to the final design doc if you need to reference it: https://docs.google.com/document/d/1T41IzREwgfqR6NVpXI8e4psfkBR39gWt0eLfvA9pkGs/edit?usp=sharing
> 
> updated animation addressing the suggested requirements. PTAL

The updates you made @nemesis09 look good to me! From the GIFs it looks like the behaviors and timing work out well.

**Comment by lwrigh** [comment]
@mceledonia Here's the PR associated with the animated quick start hints. Feel free to take a look. The GIFs at the top of the PR are the most recent version of the implementation.

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.scss
Comment (sahil143) [inline frontend/packages/console-shared/src/components/spotlight/Spotlight.scss:4]: Why didn't we use box shadow here as shown in https://codesandbox.io/s/highlight-9oizs

Comment (nemesis09) [reply frontend/packages/console-shared/src/components/spotlight/Spotlight.scss:4]: the UX for animation was updated
please find the updated UX for animation [here](https://xd.adobe.com/view/dd20d755-8def-4536-af2a-86ba83a8080c-8873/)
Slack Conversation links: 
https://coreos.slack.com/archives/C014NQG3H5L/p1604609753173100
https://coreos.slack.com/archives/C014NQG3H5L/p1604610208176400

**Comment by christianvogt** [comment]
@nemesis09 a few observations from your images:
- While the highlight border animates, a border around the item should remain in place.
- The animation spec listed here mentions nothing about the outline shrinking.
- The border should remain for 10 seconds after the animation, without fading IMO, and then fade out quickly or simply disappear. Your animation looks like a 10second slow fade out.
- Your 3rd gif shows that when the menu is scrolled, the outline fails to follow the nav item.
- I believe the guided tour outline should be the same color as the quick start highlight now.

I'm waiting on @mceledonia to provide updates on the animation before reviewing the code.

**Comment by mceledonia** [comment]
I'll run through each step of the animation and add the transitions and timings in code styling. I'll also add some visual aids for each step.

For the animation of all hint borders: 

- The user clicks the text link to activate the hint border.
- The border appears as two exact copies around the element, since they are two exact copies stacked on top of each other they just look like one. They should both `Fade in over 0.4 seconds` Right now they are static and do not move. The reason for 2 copies to appear is to achieve the desired animation effect described below.
image 
image
- After a brief `delay of 1.2 seconds` one copy of the border begins it's animation sequence. The animation sequence is as follows: `increase in scale from 100% to 150% over 1.2 seconds` `increase in border width from 4px to 8px over 1.2 seconds` `decrease in opacity/transparency from 100% to 0% over 1.2 seconds` Note - These all happen simultaneously at the same time, and finish at the same time. They do not happen one after another.
image
image
- The leftover static copy `waits/delays 10 seconds` and then triggers a `fade out over 5 seconds`
image

The spec for the starting border styling is the following:
Border width: 4px
Border color: `$pf-color-blue-200:` or `#73bcf7`

**Comment by christianvogt** [comment]
Here's an updated implementation of the requested animation:
https://codesandbox.io/s/highlight-9oizs?file=/src/Highlight.css

Note that I'm currently working with @mceledonia to tweak the animation (eg border expansion from 4px -> 12px instead of 4px -> 8px)

**Comment by christianvogt** [comment]
@nemesis09 please use the same outline for both the guided tour and quick starts highlight.

**Comment by christianvogt** [comment]
(I modified the guided tour to test this)
You need to use popperjs to position the highlight so that it moves when the node underneath it moves. Otherwise this happens:
image

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.scss
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/spotlight/Spotlight.scss:1]:
```suggestion
@keyframes co-spotlight-expand {
```

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.scss
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/spotlight/Spotlight.scss:14]:
```suggestion
@keyframes co-spotlight-fade-in {
```

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.scss
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/spotlight/Spotlight.scss:23]:
```suggestion
@keyframes co-spotlight-fade-out {
```

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx:9]: There's a bug where if `hasBackdrop === false && animate === false` then this happens:
image

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx:58]: We will need to use popperjs instead so that the spotlight moves as the node underneath it moves.

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx:33]: Clean up dead code.

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx:9]: Combine these 2 props into one called `interactive` and add a description.
If `interactive` it will allow click through and animate. If `!interactive` it will display the background and prevent click through.

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx:45]: Only listen for clicks if in `interactive` mode.
You should also make sure the click occurs on the target element and not anywhere on screen.

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.scss
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/spotlight/Spotlight.scss:36]: This causes the blue for the guided tour to be slightly a different color than the one we actually want. Perhaps raise a separate issue for this since it was an issue before as well.
image

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx
Comment (andrewballantyne) [reply frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx:45]:
```suggestion
    const handleClick = () => setClicked(true);
    element.addEventListener('mousedown', handleClick);
    return () => {
      element.removeEventListener('mousedown', handleClick);
    };
```

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx
Comment (andrewballantyne) [inline frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx:38]:
```suggestion
      element.addEventListener('mousedown', handleClick);
      return () => {
        element.removeEventListener('mousedown', handleClick);
```
This was not updated...

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.scss
Comment (nemesis09) [reply frontend/packages/console-shared/src/components/spotlight/Spotlight.scss:36]: raised issue https://issues.redhat.com/browse/ODC-5178

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.scss
Comment (nemesis09) [reply frontend/packages/console-shared/src/components/spotlight/Spotlight.scss]: done

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.scss
Comment (nemesis09) [reply frontend/packages/console-shared/src/components/spotlight/Spotlight.scss]: done

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.scss
Comment (nemesis09) [reply frontend/packages/console-shared/src/components/spotlight/Spotlight.scss]: done

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx
Comment (nemesis09) [reply frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx]: using popperjs instead of portal

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx
Comment (nemesis09) [reply frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx]: done

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx
Comment (nemesis09) [reply frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx]: done

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx
Comment (nemesis09) [reply frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx]: the goal for this function is to detect any click outside the `link` and stop the animation.

> The solid rectangle hint persists until the user clicks somewhere else on the UI or waits the 10 seconds until it fades away. 
although reading the uxd again, it seems the `click-anywhere-to-disappear` should only apply to the solid rectangle persisting after the expand animation. I'll check this with ux

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx
Comment (nemesis09) [reply frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx:38]: I'm not sure if we need to check for click event on the element, 
`element` is the component to highlight and the UXD says a click anywhere on the screen should stop the animation

Comment (christianvogt) [reply frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx:38]: I'll check with ux on the design again

Comment (christianvogt) [reply frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx:38]: I checked and you are right. They want it to disappear when clicking anywhere.

**Comment by christianvogt** [comment]
@nemesis09 this is what the guided tour now looks like:
image

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx
Comment (andrewballantyne) [inline frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx:54]: Your popper is trying to be too smart... it's pushing away from the edge of the screen.

```suggestion
    <Popper reference={element} placement="top-start" popperOptions={popperOptions}>
```

```
  import { PopperOptions } from 'popper.js';
  ...
  const popperOptions: PopperOptions = {
    modifiers: {
      preventOverflow: {
        boundariesElement: 'window',
        padding: 0,
      },
    },
  };
```

File: frontend/packages/console-shared/src/components/spotlight/__tests__/Spotlight.spec.tsx
Comment (andrewballantyne) [inline frontend/packages/console-shared/src/components/spotlight/__tests__/Spotlight.spec.tsx:24]:
```suggestion
```

File: frontend/packages/console-shared/src/components/spotlight/__tests__/Spotlight.spec.tsx
Comment (andrewballantyne) [inline frontend/packages/console-shared/src/components/spotlight/__tests__/Spotlight.spec.tsx:34]:
```suggestion
```

File: frontend/packages/console-shared/src/components/spotlight/spotlight.scss
Comment (andrewballantyne) [inline frontend/packages/console-shared/src/components/spotlight/spotlight.scss:39]:
```suggestion
```

File: frontend/packages/console-shared/src/components/spotlight/spotlight.scss
Comment (andrewballantyne) [inline frontend/packages/console-shared/src/components/spotlight/spotlight.scss:46]:
```suggestion
```

File: frontend/packages/console-shared/src/components/spotlight/InteractiveSpotlight.tsx
Comment (andrewballantyne) [inline frontend/packages/console-shared/src/components/spotlight/InteractiveSpotlight.tsx:42]:
```suggestion
    const handleClick = () => setClicked(true);
    document.addEventListener('click', handleClick);
    return () => {
      document.removeEventListener('click', handleClick);
    };
```
Changing these to `click` will help support mobile. `mousedown` is a non touch screen event.

File: frontend/packages/console-shared/src/components/spotlight/spotlight.scss
Comment (andrewballantyne) [inline frontend/packages/console-shared/src/components/spotlight/spotlight.scss:34]: I don't think this `position` is needed either since you're not specifying the `top`/`left` esk attributes.

**Review by andrewballantyne — CHANGES_REQUESTED** [review]
nits

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx
Comment (nemesis09) [reply frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx]: updated.

File: frontend/packages/console-shared/src/components/spotlight/spotlight.scss
Comment (nemesis09) [reply frontend/packages/console-shared/src/components/spotlight/spotlight.scss:34]: this is being used by guided tours as well which uses top/left.

File: frontend/packages/console-shared/src/components/spotlight/InteractiveSpotlight.tsx
Comment (nemesis09) [reply frontend/packages/console-shared/src/components/spotlight/InteractiveSpotlight.tsx]: done

File: frontend/packages/console-shared/src/components/spotlight/spotlight.scss
Comment (nemesis09) [reply frontend/packages/console-shared/src/components/spotlight/spotlight.scss]: done

File: frontend/packages/console-shared/src/components/spotlight/__tests__/Spotlight.spec.tsx
Comment (nemesis09) [reply frontend/packages/console-shared/src/components/spotlight/__tests__/Spotlight.spec.tsx]: done

File: frontend/packages/console-shared/src/components/spotlight/__tests__/Spotlight.spec.tsx
Comment (nemesis09) [reply frontend/packages/console-shared/src/components/spotlight/__tests__/Spotlight.spec.tsx]: done

**Comment by nemesis09** [comment]
> @nemesis09 this is what the guided tour now looks like:
> image

fixed

File: frontend/packages/console-shared/src/components/spotlight/spotlight.scss
Comment (andrewballantyne) [reply frontend/packages/console-shared/src/components/spotlight/spotlight.scss:34]: 🤔  Pretty sure if you don't use `top` / `left` of 0 it does wonky things... but okay. Let's leave it and worry about it if it's an issue.

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.scss
Comment (rohitkrai03) [reply frontend/packages/console-shared/src/components/spotlight/Spotlight.scss]: Shouldn't this be `ocs-spotlight-expand`?

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.scss
Comment (rohitkrai03) [reply frontend/packages/console-shared/src/components/spotlight/Spotlight.scss]: @christianvogt @nemesis09 Shouldn't this be `ocs-spotlight-expand` since this is in `console-shared` package?

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.scss
Comment (rohitkrai03) [reply frontend/packages/console-shared/src/components/spotlight/Spotlight.scss]: @christianvogt @nemesis09 Shouldn't this be `ocs-spotlight-expand` since this is in `console-shared` package?

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.scss
Comment (andrewballantyne) [reply frontend/packages/console-shared/src/components/spotlight/Spotlight.scss]: Not sure it matters... we have a poor consistency across the console for keyframes and their names having the appropriate module. I think it's because they are only ever used in the one file. I'm indifferent either way though.

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.scss
Comment (christianvogt) [reply frontend/packages/console-shared/src/components/spotlight/Spotlight.scss]: yes it should be. my mistake

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.scss
Comment (christianvogt) [reply frontend/packages/console-shared/src/components/spotlight/Spotlight.scss]: yes it should be. my mistake

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.scss
Comment (christianvogt) [reply frontend/packages/console-shared/src/components/spotlight/Spotlight.scss]: yes it should be. my mistake

**Review by andrewballantyne — APPROVED** [review]
/lgtm

/assign @christianvogt  @rohitkrai03 

Need an approval here when one of you get a chance.

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.scss
Comment (nemesis09) [reply frontend/packages/console-shared/src/components/spotlight/Spotlight.scss]: renamed to `ocs-spotlight-expand`

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.scss
Comment (nemesis09) [reply frontend/packages/console-shared/src/components/spotlight/Spotlight.scss]: enamed to ocs-spotlight-expand

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.scss
Comment (nemesis09) [reply frontend/packages/console-shared/src/components/spotlight/Spotlight.scss]: renamed to `ocs-spotlight-expand`

**Comment by andrewballantyne** [comment]
Odd, pod didn't start...

> * 1x kubelet: Successfully pulled image "quay.io/prometheus/busybox:latest" in 1m20.65114373s
2020/11/30 16:34:27 pod didn't start running within 15m0s: 
> * Container artifacts is not ready with reason ContainerCreating
> * Container release is not ready with reason ContainerCreating

/retest

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]
@nemesis09 I was testing this with guided tours by passing interactive prop to spotlight. I noticed some weird behaviors.

- The outside click doesn't work on any part of the document. It only works when clicked on the popper part of the guided tour.
- The borders on the help menu looks off. It looks okay without the `interactive` prop passed but the arrow is still off a bit from the center.
 
Screenshot from 2020-11-30 22-30-54

Screenshot from 2020-11-30 22-29-21

**Comment by christianvogt** [comment]
Interesting that the popover for help is also offset.

**Comment by andrewballantyne** [comment]
I don't understand how others are reproducing this situation lol... Christian on his Mac can reproduce... Aritra on Linux and Rohit on Linux -- but I can't on my Mac lol

Chrome
Screen Shot 2020-11-30 at 1 57 39 PM

FF
Screen Shot 2020-11-30 at 1 58 12 PM

**Comment by andrewballantyne** [comment]
> The outside click doesn't work on any part of the document. It only works when clicked on the popper part of the guided tour.

@rohitkrai03 are you saying during Guided Tour you are expecting to click on the mask and exit? Because the interactive mode is false and thus the only way to close the modal is to close it via the `x` or finish the flow (or use one the exit at the start)

**Comment by nemesis09** [comment]
> @nemesis09 I was testing this with guided tours by passing interactive prop to spotlight. I noticed some weird behaviors.
> 
>     * The outside click doesn't work on any part of the document. It only works when clicked on the popper part of the guided tour.
> 
>     * The borders on the help menu looks off. It looks okay without the `interactive` prop passed but the arrow is still off a bit from the center.
> 
> 
> Screenshot from 2020-11-30 22-30-54
> 
> Screenshot from 2020-11-30 22-29-21

This is odd. As far as I understand, there has been no substantial change in the spotlight as far as the implementation for Guided Tour is concerned.

**Comment by andrewballantyne** [comment]
> This is odd. As far as I understand, there has been no substantial change in the spotlight as far as the implementation for Guided Tour is concerned.

@rohitkrai03 / @christianvogt / @nemesis09  Can you fire up any cluster (before this is merged) and check the Guided Tour? Is this a pre-existing issue? I obviously can't reproduce this 😞

**Comment by nemesis09** [comment]
> I don't understand how others are reproducing this situation lol... Christian on his Mac can reproduce... Aritra on Linux and Rohit on Linux -- but I can't on my Mac lol
> 
> Chrome
> Screen Shot 2020-11-30 at 1 57 39 PM
> 
> FF
> Screen Shot 2020-11-30 at 1 58 12 PM

Update: I tried it out in the cluster, rather than localhost to check the differences. The offset in the arrow seems to be a pre-existing issue. I think this could be related to a similar behaviour in popover arrow that i noticed in the modals for cluster status card items, which later turned out to be a patternfly issue. here is the slack thread https://coreos.slack.com/archives/C6A3NV5J9/p1603962222473200

**Comment by nemesis09** [comment]
> @nemesis09 I was testing this with guided tours by passing interactive prop to spotlight. I noticed some weird behaviors.
> 
>     * The outside click doesn't work on any part of the document. It only works when clicked on the popper part of the guided tour.
> 
>     * The borders on the help menu looks off. It looks okay without the `interactive` prop passed but the arrow is still off a bit from the center.
> 
> 
> Screenshot from 2020-11-30 22-30-54
> 
> Screenshot from 2020-11-30 22-29-21

Update: I tried it out in the cluster, rather than localhost to check the differences. The offset in the arrow seems to be a pre-existing issue. I think this could be related to a similar behaviour in popover arrow that i noticed in the modals for cluster status card items, which later turned out to be a patternfly issue. here is the slack thread https://coreos.slack.com/archives/C6A3NV5J9/p1603962222473200

**Comment by rohitkrai03** [comment]
> > The outside click doesn't work on any part of the document. It only works when clicked on the popper part of the guided tour.
> 
> @rohitkrai03 are you saying during Guided Tour you are expecting to click on the mask and exit? Because the interactive mode is false and thus the only way to close the modal is to close it via the `x` or finish the flow (or use one the exit at the start)

@andrewballantyne I tested guided tour after passing `interactive` prop to the `Spotlight` component. So my expectation was that clicking anywhere should stop the animation and fade out the border but it only works when clicked on the popover.

The below screenshot shows the original use of `Spotlight` without interactive prop. The arrow seems to be offset which is a pre-existing issue even before this PR but the border box around help is correctly placed.
Screenshot from 2020-11-30 22-30-54

The below screenshot shows the `Spotlight` with `interactive` prop passed so that it uses `InteractiveSpotlight` with animation. It has the border box around help icon offset to the icon but the arrow is now centered with the border box.
Screenshot from 2020-11-30 22-29-21

**Comment by andrewballantyne** [comment]
> @andrewballantyne I tested guided tour after passing `interactive` prop to the `Spotlight` component. So my expectation was that clicking anywhere should stop the animation and fade out the border but it only works when clicked on the popover

Well that's disconcerting... I'll have to test this out once I am finished with my current task.


> The below screenshot shows the original use of `Spotlight` without interactive prop. The arrow seems to be offset which is a pre-existing issue even before this PR but the border box around help is correctly placed.
> Screenshot from 2020-11-30 22-30-54
> 
> The below screenshot shows the `Spotlight` with `interactive` prop passed so that it uses `InteractiveSpotlight` with animation. It has the border box around help icon offset to the icon but the arrow is now centered with the border box.
> Screenshot from 2020-11-30 22-29-21

This is very confusing lol... I can't reproduce this in any fashion... I will have to try all the browsers under my disposal to see if I can reproduce this. Either way, I think we are addressing this post this PR merge.

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx:11]:
```suggestion
  const element = React.useMemo(() => document.querySelector(selector), [selector]);
```

File: frontend/packages/console-shared/src/components/spotlight/InteractiveSpotlight.tsx
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/spotlight/InteractiveSpotlight.tsx:31]: This is a utility function. Move it out of the render closure.

File: frontend/packages/console-shared/src/components/spotlight/InteractiveSpotlight.tsx
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/spotlight/InteractiveSpotlight.tsx:22]: Make this a const outside of the render function.

File: frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx
Comment (nemesis09) [reply frontend/packages/console-shared/src/components/spotlight/Spotlight.tsx]: done

File: frontend/packages/console-shared/src/components/spotlight/InteractiveSpotlight.tsx
Comment (nemesis09) [reply frontend/packages/console-shared/src/components/spotlight/InteractiveSpotlight.tsx]: done

File: frontend/packages/console-shared/src/components/spotlight/InteractiveSpotlight.tsx
Comment (nemesis09) [reply frontend/packages/console-shared/src/components/spotlight/InteractiveSpotlight.tsx]: done

**Review by andrewballantyne — APPROVED** [review]
/lgtm

**Pull request merged**
